### DISCUSSION TITLE
how can i define a custom metrics in oal?

### DISCUSSION THREAD

**Discussion by tjzheng1002** [apache/skywalking#10403] [Q&A]
I define a custom metrics app_error_cpm,but it happened error：
```
line 100:21 mismatched input 'AppError' expecting {'Service', 'TCPService', 'ServiceInstance', 'TCPServiceInstance', 'Endpoint', 'ServiceRelation', 'TCPServiceRelation', 'ServiceInstanceRelation', 'TCPServiceInstanceRelation', 'EndpointRelation', 'ServiceInstanceJVMCPU', 'ServiceInstanceJVMMemory', 'ServiceInstanceJVMMemoryPool', 'ServiceInstanceJVMGC', 'ServiceInstanceJVMThread', 'ServiceInstanceJVMClass', 'DatabaseAccess', 'ServiceInstanceCLRCPU', 'ServiceInstanceCLRGC', 'ServiceInstanceCLRThread', 'EnvoyInstanceMetric', 'Event', 'CacheAccess', 'MQAccess', 'MQEndpointAccess', 'BrowserAppPerf', 'BrowserAppPagePerf', 'BrowserAppSingleVersionPerf', 'BrowserAppTraffic', 'BrowserAppPageTraffic', 'BrowserAppSingleVersionTraffic', '(str->long)', '(long)', '(str->int)', '(int)'}
line 100:29 mismatched input '.' expecting {'=', SPACE}
line 100:36 mismatched input '(' expecting {'=', SPACE}
```
My modification is as follows：  
**core.oal：**
`app_error_cpm = from(AppError.*).cpm();`
**DefaultScopeDefine：**
```
public static final int APP_ERROR = 314;
public static final String APP_ERROR_CATALOG_NAME = "APP_ERROR";
```
**AppError**
```
@ScopeDeclaration(id = APP_ERROR, name = "AppError", catalog = APP_ERROR_CATALOG_NAME)
@ScopeDefaultColumn.VirtualColumnDefinition(fieldName = "entityId", columnName = "entity_id", isID = true, type = String.class)
@Slf4j
public class AppError extends Source {
    private String entityId;

    @Override
    public int scope() {
        return DefaultScopeDefine.APP_ERROR;
    }
```
**OALLexer.g4**
```
// App Error
SRC_APP_ERROR: 'AppError';
```
**OALParser.g4**
```
source
    : SRC_SERVICE | SRC_TCP_SERVICE | SRC_DATABASE_ACCESS | SRC_SERVICE_INSTANCE | SRC_TCP_SERVICE_INSTANCE | SRC_ENDPOINT | SRC_CACHE_ACCESS |
......
      SRC_EVENT | SRC_MQ_ACCESS | SRC_MQ_ENDPOINT_ACCESS | SRC_APP_ERROR
    ;
```
Any idea aboat this question?Any learning aboat oal can provide?I have no idea now

**Comment by wu-sheng** [comment]
If you want to extend source, you have to change g4 files to regenerate grammar tree.
Usually we don't recommend this, too complex, and community is hard to follow your private changes. Mostly, we have to be on your own.

**Reply by tjzheng1002** [reply]
> If you want to extend source, you have to change g4 files to regenerate grammar tree. Usually we don't recommend this, too complex, and community is hard to follow your private changes. Mostly, we have to be on your own.

Thank you for your answer @wu-sheng . I repackage the project so as not to report the error. The oal-grammar error was not processed by packing it separately before, so run it directly and report that error.

**Reply by wu-sheng** [reply]
If you want to understand, learn antlr and how g4 and OAL compiling work.

**Reply by tjzheng1002** [reply]
> If you want to understand, learn antlr and how g4 and OAL compiling work.

Okay, I'll check it out!
